specifyweb/specify/tree_ranks.py

```diff
@@ -0,0 +1,277 @@
+TREE_RANKS_MAPPING = {
+    'taxon': (TAXON_RANKS, TAXON_RANK_INCREMENT),
+    'geography': (GEOGRAPHY_RANKS, GEOGRAPHY_RANK_INCREMENT),
+    'storage': (STORAGE_RANKS, STORAGE_RANK_INCREMENT),
+    'geologictimeperiod': (GEOLOGIC_TIME_PERIOD_RANKS, GEOLOGIC_TIME_PERIOD_INCREMENT),
+    'lithostrat': (LITHO_STRAT_RANKS, LITHO_STRAT_INCREMENT),
+}
+
+def post_tree_rank_save(tree_def_item_model, new_rank):
+    tree_def = new_rank.treedef
+    parent_rank = new_rank.parent
+    new_rank_id = new_rank.rankid
+
+    # Set the parent rank, that previously pointed to the target, to the new rank
+    child_ranks = tree_def_item_model.objects.filter(treedef=tree_def, parent=parent_rank).exclude(rankid=new_rank_id)
+    if child_ranks.exists():
+        # Iterate through the child ranks, but there should only ever be 0 or 1 child ranks to update
+        for child_rank in child_ranks:
+            child_rank.parent = new_rank
+            child_rank.save()
+
+    # Update the old root rank to point to the new root rank if the new rank is the new root rank
+    if new_rank.parent is None:
+        old_root_rank = tree_def_item_model.objects.exclude(id=new_rank.id).filter(treedef=tree_def, parent=None).first()
+        if old_root_rank is not None:
+            old_root_rank.parent = new_rank
+            old_root_rank.save()
+
+    # Regenerate full names
+    tree_extras.set_fullnames(tree_def, null_only=False, node_number_range=None)
+
+def pre_tree_rank_deletion(tree_def_item_model, rank):
+    tree_def = rank.treedef
+
+    # Make sure no nodes are present in the rank before deleting rank
+    if tree_def_item_model.objects.filter(parent=rank).count() > 1:
+        raise TreeBusinessRuleException("The Rank {rank.name} is not empty, cannot delete!")
+
+    # Set the parent rank, that previously pointed to the old rank, to the target rank
+    child_ranks = tree_def_item_model.objects.filter(treedef=tree_def, parent=rank)
+    if child_ranks.exists():
+        # Iterate through the child ranks, but there should only ever be 0 or 1 child ranks to update
+        for child_rank in child_ranks:
+            child_rank.parent = rank.parent
+            child_rank.save()
+            print(child_rank.name)
+
+def post_tree_rank_deletion(rank):
+    # Regenerate full names
+    tree_extras.set_fullnames(rank.treedef, null_only=False, node_number_range=None)
+
+def pre_tree_rank_init(new_rank):
+    set_rank_id(new_rank)
+
+def set_rank_id(new_rank):
+    """
+    Sets the new rank to the specified tree when adding a new rank.
+    Expects at least the name and parent of the rank to be set.
+    All the other parameters are optional.
+    """
+    # Get parameter values from data
+    tree = new_rank.specify_model.name.replace("TreeDefItem", "").lower()
+    new_rank_name = getattr(new_rank, 'name', None)
+    parent_rank_name = getattr(new_rank.parent, 'name', 'root') if getattr(new_rank, 'parent', None) else 'root'
```

Comment:
can we make sure that editing the parent rank name does not break the linking of the child rank?
do we need to have a business rule for this?